File: MVVMJavaRx/app/src/main/java/com/vascome/fogtail/presentation/base/fragments/BaseFragment.java
package com.vascome.fogtail.presentation.base.fragments;

import android.support.annotation.NonNull;

import com.vascome.fogtail.utils.LeakCanaryProxy;

import javax.inject.Inject;

import dagger.android.support.DaggerFragment;
import io.reactivex.disposables.CompositeDisposable;


/**
 * Created by vasilypopov on 11/23/17
 * Copyright (c) 2017 MVPJava. All rights reserved.
 */

public abstract class BaseFragment extends DaggerFragment {

    @Inject
    LeakCanaryProxy leakCanaryProxy;

    protected CompositeDisposable disposables = new CompositeDisposable();

    protected void runIfFragmentAlive(@NonNull Runnable runnable) {
        if (isFragmentAlive()) {
            runnable.run();
        }
    }

    private boolean isFragmentAlive() {
        return getActivity() != null && isAdded() && !isDetached() && getView() != null && !isRemoving();
    }

    @Override
    public void onPause() {
        if (!disposables.isDisposed()) {
            disposables.dispose();
        }
        super.onPause();
    }

    @Override
    public void onResume() {
        disposables = new CompositeDisposable();
        super.onResume();
    }

    @Override
    public void onDestroy() {
        leakCanaryProxy.watch(this);
        super.onDestroy();
    }
}

File: MVVMKotlinArchComponent/app/src/main/java/com/vascome/fogtail/presentation/base/fragments/BaseFragment.kt
package com.vascome.fogtail.presentation.base.fragments

import com.vascome.fogtail.utils.LeakCanaryProxy
import dagger.android.support.DaggerFragment
import io.reactivex.disposables.CompositeDisposable
import javax.inject.Inject


/**
 * Created by vasilypopov on 11/23/17
 * Copyright (c) 2017 MVPJava. All rights reserved.
 */

abstract class BaseFragment : DaggerFragment() {

    @Inject
    lateinit var leakCanaryProxy: LeakCanaryProxy

    protected var disposables = CompositeDisposable()

    private val isFragmentAlive: Boolean
        get() = activity != null && isAdded && !isDetached && view != null && !isRemoving

    protected fun runIfFragmentAlive(runnable: Runnable) {
        if (isFragmentAlive) {
            runnable.run()
        }
    }

    override fun onPause() {
        if (!disposables.isDisposed) {
            disposables.dispose()
        }
        super.onPause()
    }

    override fun onResume() {
        disposables = CompositeDisposable()
        super.onResume()
    }

    override fun onDestroy() {
        leakCanaryProxy.watch(this)
        super.onDestroy()
    }
}
